### ISSUE TITLE
Option to warn about incompatible phpdoc types?

### ISSUE THREAD

**Issue by GrahamCampbell** [vimeo/psalm#3785]
How do I get psalm to warn me about the phpdoc type not including null:

```php
/**
 * @param string $x
 */
function f(string $x = null) {}
```

`string` is not correct phpdoc, but `?string` or `string|null` would be.

Simiarly, if there is no type on the method, one should still be able to detect there is a problem. This phpdoc is also wrong:

```php
/**
 * @param string $x
 */
function f($x = null) {}
```

**Comment by GrahamCampbell** [comment]
> can you reproduce the issue on https://psalm.dev

Yes, no warning raised.

**Comment by muglug** [comment]
Psalm will automatically add `null` to docblock param types when it sees `null` in the signature (either as a default or in the type itself).

It's a way to avoid things that probably aren't _bugs_ per se, just lazy documentation.

**Comment by GrahamCampbell** [comment]
But how do I get it to report it as an error? I want my code have correct type documentation for human consumers.

**Comment by muglug** [comment]
> I want my code have correct type documentation for human consumers.

This happens during reflection in three places:
- https://github.com/vimeo/psalm/blob/0119cd09c11ea1630d4fe2c99f9b5c52d8063272/src/Psalm/Internal/PhpVisitor/ReflectorVisitor.php#L2774
- https://github.com/vimeo/psalm/blob/0119cd09c11ea1630d4fe2c99f9b5c52d8063272/src/Psalm/Internal/PhpVisitor/ReflectorVisitor.php#L3244
- https://github.com/vimeo/psalm/blob/0119cd09c11ea1630d4fe2c99f9b5c52d8063272/src/Psalm/Internal/PhpVisitor/ReflectorVisitor.php#L3485

You could add some sort of config flag (off by default so existing behaviour stays the same) to emit an issue whenever that happens.

**Comment by GrahamCampbell** [comment]
I was thinking this would be useful for lots of people, and that it should be added in the core.

**Comment by muglug** [comment]
> I was thinking this would be useful for lots of people

Yeah, but I think it would be annoying for a much greater number – when adding static analysis to Vimeo's codebase, I saw a lot of these errors initially, then thought "hmm that's something the tool can work around" (and it seems like Ondrej has the same approach in PHPStan).

I agree that for human consumers it might be slightly incomplete, but that's a code style problem, not one that could lead to incorrect type inference or bugs.

**Comment by muglug** [comment]
So to be clear, I'm happy to have it in core, but not turned on by default.

**Comment by muglug** [comment]
Reopening this in case you want to create a PR (let me know if you need any more pointers).

**Comment by orklah** [comment]
While I'd not be very much interested by having the error for itself, having a psalter fix for that would be great!

**Comment by mr-feek** [comment]
I feel like I've seen something similar in phpcs before? Maybe something in https://github.com/slevomat/coding-standard#slevomatcodingstandardtypehintsparametertypehint- ? Perhaps I'm thinking of phpstan though..

**Comment by orklah** [comment]
If anyone still looking for something like that, I made a quick plugin to add the "?" for nullable type with Psalter:
https://gist.github.com/orklah/559c6eee0d17f94645755fb7256c8fc1

If there's interest, I can publish it under a composer package. I could probably extend it to affect phpdoc too.

**Comment by orklah** [comment]
I'll close this as this is indeed something more suited for a CS tool. Still happy to accept a PR if someone wants to propose it.
